### PR TITLE
Adding support for X type, updated NMAP, README, and added RADIO cont

### PR DESCRIPTION
**Pull request by brytul**
X types can now be used with the writeNewActionFlows() function.  X types doesnt take the actionset yet.

### CODE REVIEW

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/middleboxmgmnt/AlertHandler.java [hunk 1]
```diff
@@ -118,14 +121,23 @@ public void run() {
...
 		    // Alert Msg Analysis
-		    MsgAnalysis analyzer = new MsgAnalysis(alert, devPolicy[Integer.parseInt(policyID)].getTransition()[policyMap.get(srcMac).getStateNum()]);
+
+		    //MsgAnalysis analyzer = new MsgAnalysis(alert, devPolicy[Integer.parseInt(policyID)].getTransition()[policyMap.get(srcMac).getStateNum()]);
+		    MsgAnalysis analyzer = new MsgAnalysis(alert, devPolicy[Integer.parseInt(policyID)], this.policyMap, srcMac);
+
 		    if(analyzer.analyze()) {
+		    System.out.println("next state!");
+		    int timer = 5;
+		    while (timer!=0){
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/middleboxmgmnt/AlertHandler.java]: what is this timer for?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/FlowWriterServiceImpl.java [hunk 2]
```diff
@@ -220,7 +220,7 @@ public void addBidirectionalMacToMacFlows(MacAddress sourceMac,
     }
 
     @Override
-    public void addBidirectionalMacFlows(MacAddress sourceMac,
+    public void addBidirectionalMacFlows(MacAddress sourceMac, //nmap
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/FlowWriterServiceImpl.java]: I'm not sure what this comment is for?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/ReactiveFlowWriter.java [hunk 3]
```diff
@@ -191,9 +200,9 @@ public void writeFlows(NodeConnectorRef ingress, MacAddress srcMac, NodeConnecto
         }
     }
 
-    public void writeFlows(RuleDescriptor rule){
+    public void writeFlows(RuleDescriptor rule){ //nmap [now we use line writeNewActionFlows(rule)]
 	if(rule.outMac.equals("*")){
-	    flowWriterService.addBidirectionalMacFlows(new MacAddress(rule.inMac), rule.inNCR, rule.outNCR);
+	    flowWriterService.addBidirectionalMacFlows(new MacAddress(rule.inMac), rule.inNCR, rule.outNCR); //nmap
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/ReactiveFlowWriter.java]: these are not used by the nmap container anymore

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java [hunk 4]
```diff
@@ -62,8 +65,35 @@ public ServiceChain(String dataplaneIP, String dockerPort, String ovsPort,
...
+	}
+	this.protectDetails=devPolicy.getProtections()[i];
+	this.curState=devPolicy.getStates()[i];
+	this.devNum=devNum;
+	this.nodeStr=this.containerCalls.getNodeString(ncr);
+	this.inNCR=inNCR;
+	this.outNCR=outNCR;
+	this.srcMac = srcMac;
+    } // adds srcMac for cookie conversion    
+
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: Is this used?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java [hunk 5]
```diff
@@ -155,8 +185,8 @@ private void restartContProcess(String contName) {
 	return ncrs;
     }
     
-    public NodeConnectorRef[] startAccessibleCont_getNCR(String contName, String contImage, String[] ifaces, String ip) {
-	this.containerCalls.startContainer(contName, contImage, this.devNum);
+    public NodeConnectorRef[] startAccessibleCont_getNCR(String contName, String contImage, String[] ifaces, String ip, String iot_IP) { //nmap
+	this.containerCalls.startContainer(contName, contImage, this.devNum, iot_IP);
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: so do you have to provide the target IP address now?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java [hunk 6]
```diff
@@ -254,13 +284,13 @@ public NewFlows setupChain() {
 		    nodes.add(newNode);
 		}
 	    }
-	    else if(chainLinks[i].equals("A")){
+	    else if(chainLinks[i].equals("A")){ //addressable proxy
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: Should this simply be if A or X?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java [hunk 7]
```diff
@@ -277,9 +307,48 @@ else if(chainLinks[i].equals("A")){
...
+
+			enableARPs(protectDetails.imageOpts[i].contName, ifaces, inNCR, outNCR);
+			contMac = getContMacAddress(protectDetails.imageOpts[i].contName, ifaces[0]);
+			MacGroup newGroupA = new MacGroup(inMac, contMac.getValue());
+			MacGroup newGroupB = new MacGroup(contMac.getValue(), outMac);
+			groups.add(newGroupA);
+			groups.add(newGroupB);
+			macMap.put(groupCnt, i);
+			groupCnt++;
+	    }
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: Is this different from the if A?

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java [hunk 8]
```diff
@@ -277,9 +307,48 @@ else if(chainLinks[i].equals("A")){
...
+		System.out.println("X cont detected - adding inNCR to nodes");
+		nodes.add(inNCR);
+	}else{
+		nodes.add(outNCR);
+	}
+	for (NodeConnectorRef a_node:nodes){
+		String[] findOFport = a_node.getValue().firstKeyOf(NodeConnector.class).getId().getValue().split(":");
+		System.out.println("A node: " +findOFport[findOFport.length-1]);
+
+	}
```
Comment (slab14) [inline l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: is this just for debugging?

File: l2switch-main/impl/src/main/yang/l2switch-config.yang [hunk 9]
```diff
@@ -54,7 +54,7 @@ module l2switch-config {
         }
 	leaf policy-file {
 	     type string;
-	     default "/users/slab/IoT_Sec_Gateway/policies/cloudlab-NewPolicy20.json";
+	     default "/users/brytul/IoT_Sec_Gateway/policies/nmap.json";
```
Comment (slab14) [inline l2switch-main/impl/src/main/yang/l2switch-config.yang]: lets eventually move this to /etc/sec_gateway/policy/<name>.json

**Review by slab14 — COMMENTED** [review]
Cool work in progress. Let's get a full end-to-end implementation before merging upstream.

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java (hunk 5 above)
Comment (brytul) [reply l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: for both X and A type containers, yes.  It's gathered by the ARP packet handler

File: l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java (hunk 8 above)
Comment (brytul) [reply l2switch-main/impl/src/main/java/org/opendaylight/l2switch/flow/chains/ServiceChain.java]: yes, its removed

**Comment by brytul** [comment]
NMAP impl finished end-to-end

**Pull request merged**